GVRf/Framework/src/org/gearvrf/utility/Preconditions.java

```diff
@@ -0,0 +1,103 @@
+    public static final String INPUT_TYPE_TEXCOORDS = "Input texture coordinates";
+
+    /**
+     * A simple check to see if the object is non-null or otherwise the function will throw an
+     *  IllegalArgumentException.
+     *  
+     * @param object
+     *          the parameter object can be any instance of any class even an array.
+     * @param dataMessage
+     *          To point out the type of data, for the use when we try to throw an exception.
+     *          Type may not refer to the class name but a wider concept of class type such as vertices, normals.
+     * @throws IllegalArgumentException if the object is null pointer.
+     */
+    public static void checkNotNull(Object object, String dataMessage) {
+        if (object == null) {
+            throw Exceptions.IllegalArgument("%s can't be a null pointer.",
+                    dataMessage);
+        }
+    }
+    
+    /**
+     * We can see that there is a wide use of key-value pair for the parameters which are used in shader.
+     * This function is to ensure that the key is meaningful.
+     * @param key
+     *          Key in the key-value pair.
+     *          
+     * @throws IllegalArgumentException if the key is null or empty.
+     */
+    public static void checkKeyNotNullOrEmpty(String key){
+        if(key == null || key.trim().isEmpty()){
+            throw Exceptions.IllegalArgument("Current key is %s which is not a proper key.", key == null ? "null" : "empty");
+        }
+    }
+
+    /**
+     * In GVRF, there are a lot of arrays uses with expected number of components inside.
+     * This function is to ensure that the input array is legal according to the expected components' number. 
+     * Otherwise it will throw an IllegalArgumentException.
+     * @param dataLength
+     *          Length of array data.
+     * @param expectedComponents
+     *          Expected number of components.
+     * @throws IllegalArgumentException if the dataLength is not divisible by the expected number.
+     */         
+    public static void checkDivisibleDataLength(int dataLength, int expectedComponents) {
+        if (dataLength == 0 || dataLength % expectedComponents != 0) {
+            throw Exceptions
+                    .IllegalArgument(
+                            "The input array should be an array of %d-component elements whose length is non-zero and divisible by %d. But current data length is %d.",
+                            expectedComponents, expectedComponents, dataLength);
+        }
+    }
+    
+    /**
+     * In common shader cases, NaN makes little sense. Correspondingly, GVRF is going to use Float.NaN as illegal 
+     * flag in many cases. Therefore, we need a function to check if there is any setX that is using NaN as input.
+     * @param data
+     *         A series of input float data.
+     * @throws IllegalArgumentException if the data includes NaN.
+     */
+    public static void checkNotNaN(float ...data){
```

Comment:
A `float...` parameter with only one parameter generates code like

    float[] __temporary__ = new float[1];
    __temporary__[0] = parameter;

... and is thus comparatively expensive. (`calloc()` costs, plus making the next garbage collection happen that much sooner). It's fine to offer this method ... but please also offer an overload that takes a single float parameter.